weblab/datasets/views.py

```diff
@@ -230,3 +231,48 @@ def test_func(self):
     def get_success_url(self, *args, **kwargs):
         ns = self.request.resolver_match.namespace
         return reverse(ns + ':list')
+
+
+class DatasetTransferView(LoginRequiredMixin, UserFormKwargsMixin, UserPassesTestMixin,
+                          FormMixin, DetailView):
+    template_name = 'datasets/dataset_transfer_ownership.html'
+    context_object_name = 'dataset'
+    form_class = DatasetTransferForm
+    model = Dataset
+
+    def _get_object(self):
+        if not hasattr(self, 'object'):
+            self.object = self.get_object()
+        return self.object
+
+    def test_func(self):
+        return self._get_object().is_managed_by(self.request.user)
+
+    def post(self, request, *args, **kwargs):
+        """Check the form and transfer ownership of the entity.
+
+        Called by Django when a form is submitted.
+        """
+        form = self.get_form()
+
+        if form.is_valid():
+            user = form.cleaned_data['user']
+            dataset = self.get_object()
+            if self.model.objects.filter(name=dataset.name, author=user).exists():
+                form.add_error(None, "User already has a dataset called %s" % dataset.name)
+                return self.form_invalid(form)
+
+            old_path = dataset.archive_path
+            dataset.author = user
+            dataset.save()
+            new_path = dataset.archive_path
+            if old_path.exists():
+                shutil.move(str(old_path), str(new_path))
```

Comment:
I've figured it out! The error is actually complaining that the *destination folder* doesn't exist, not the source file. So we need to create that folder first (and indeed delete the source folder after doing the move) using the same approach as in `datasets/signals.py` for when a dataset is created or deleted:
```suggestion
            dataset.abs_path.mkdir(exist_ok=False, parents=True)
            if old_path.exists():
                shutil.move(str(old_path), str(new_path))
            if old_path.parent.is_dir():
                shutil.rmtree(str(old_path.parent))
```